File: 2104-sum-of-subarray-ranges/2104-sum-of-subarray-ranges.cpp
// @Author: KING-SEN

class Solution {
public:
    long long subArrayRanges(vector<int>& nums) {
        int n = nums.size();
        long long answer = 0;
        // Decreasing stack => finds the previous Greater and next greater elements
        stack<int> stack;
        for (int right = 0; right <= n; right++) {
            while (!stack.empty() && (right == n || nums[stack.top()] <= nums[right])) {
                int mid = stack.top();
                stack.pop();
                int left = stack.empty() ? -1 : stack.top();
                answer += 1ll * nums[mid] * (mid - left) * (right - mid);
            }
            stack.push(right);
        }
        
        // Increasing stack => finds the previous Lesser and next Lesser elements
        stack.pop();
        for (int right = 0; right <= n; right++) {
            while (!stack.empty() && (right == n || nums[stack.top()] >= nums[right])) {
                int mid = stack.top();
                stack.pop();
                int left = stack.empty() ? -1 : stack.top();
                answer -= 1ll* nums[mid] * (mid - left) * (right - mid);
            }
            stack.push(right);
        }
        return answer;
    }
};

File: 2104-sum-of-subarray-ranges/2104-sum-of-subarray-ranges.py
# KING-SEN

class Solution:
    def subArrayRanges(self, nums: List[int]) -> int:
        n = len(nums)
        res = 0
        stack = []
        # Decresing stack for maximum element in the range
        for right in range(n + 1):
            while stack and (right == n or nums[stack[-1]] <= nums[right]):
                j = stack.pop()
                left = -1 if not stack else stack[-1]
                res += (j - left) * (right - j) * nums[j]
            stack.append(right)
        stack = []
        # Increasing stack for minimum element in the range
        for right in range(n + 1):
            while stack and (right == n or nums[stack[-1]] >= nums[right]):
                j = stack.pop()
                left = -1 if not stack else stack[-1]
                res -= (j - left) * (right - j) * nums[j]
            stack.append(right)
        return res